draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1958 @@
+  * supports implementation of Unix/POSIX command line conventions
+  * supports loading values from environment variables out of the box
+  * supports for prompting of custom values
+  * is fully nestable and composable
+  * works the same in Python 2 and 3
+  * supports file handling out of the box
+  * comes with useful common helpers (getting terminal dimensions,
+    ANSI colors, fetching direct keyboard input, screen clearing,
+    finding config paths, launching apps and editors, etc.)
+
+  There are many alternatives to Click and you can have a look at them if
+  you enjoy them better.  The obvious ones are ``optparse`` and ``argparse``
+  from the standard library.
+
+  Click actually implements its own parsing of arguments and does not use
+  ``optparse`` or ``argparse`` following the ``optparse`` parsing behavior.
+  The reason it's not based on ``argparse`` is that ``argparse`` does not
+  allow proper nesting of commands by design and has some deficiencies when
+  it comes to POSIX compliant argument handling.
+
+  Click is designed to be fun to work with and at the same time not stand in
+  your way.  It's not overly flexible either.  Currently, for instance, it
+  does not allow you to customize the help pages too much. This is intentional
+  because Click is designed to allow you to nest command line utilities.  The
+  idea is that you can have a system that works together with another system by
+  tacking two Click instances together and they will continue working as they
+  should.
+
+  Too much customizability would break this promise.
+
+Click describes it's
+`advantages over argparse <https://click.palletsprojects.com/en/7.x/why/#why-not-argparse>`_
+in its documentation as well:
+
+  Click is internally based on optparse instead of argparse.  This however
+  is an implementation detail that a user does not have to be concerned
+  with.  The reason however Click is not using argparse is that it has some
+  problematic behaviors that make handling arbitrary command line interfaces
+  hard:
+
+  * argparse has built-in magic behavior to guess if something is an
+    argument or an option.  This becomes a problem when dealing with
+    incomplete command lines as it's not possible to know without having a
+    full understanding of the command line how the parser is going to
+    behave.  This goes against Click's ambitions of dispatching to
+    subparsers.
+  * argparse currently does not support disabling of interspersed
+    arguments.  Without this feature it's not possible to safely implement
+    Click's nested parsing nature.
+
+In contrast to :mod:`argparse`, the `Click community <https://github.com/click-contrib>`_
+provides many extensions we can use to create a better user experience
+for our users.
+
+Click supports calling `async` methods and functions
+using the `trio-click <https://github.com/click-contrib/trio-click>`_ extension
+which is likely to be important for us in the future.
```

Comment:
@thedrow this project was renamed to https://github.com/click-contrib/asyncclick and supports asyncio and trio with anyio